.github/workflows/cli.yml

```diff
@@ -43,28 +50,445 @@ jobs:
+          azdev setup -r .
+          azdev extension add apic-extension
+
+      - name: Run tests
+        run: |
+          cd azure-cli-extensions
+          source .venv/bin/activate
+          if [ "${{ matrix.test-case }}" = "full-suite" ]; then
+            echo "Running full test suite..."
+            azdev test apic-extension --discover --live
+          else
+            echo "Running individual test case: ${{ matrix.test-case }}"
+            azdev test azext_apic-extension ${{ matrix.test-case }} --discover --live
+          fi
+        env:
+          USERASSIGNED_IDENTITY: ${{ secrets.USERASSIGNED_IDENTITY }}
+          AWS_ACCESS_KEY_LINK: ${{ secrets.AWS_ACCESS_KEY_LINK }}
+          AWS_SECRET_ACCESS_KEY_LINK: ${{ secrets.AWS_SECRET_ACCESS_KEY_LINK }}
+  execute-test-case-py311:
+    needs: setup
+    runs-on: ubuntu-latest
+    environment: engineering
+    strategy:
+      fail-fast: false
+      matrix:
+        # Use discovered test cases if available, otherwise use a single placeholder
+        test-case: ${{ fromJson(needs.setup.outputs.test-cases != '[]' && needs.setup.outputs.test-cases || '["full-suite"]') }}
+    name: ${{ matrix.test-case == 'full-suite' && 'Run Full Test Suite (Python 3.11)' || format('{0} (Python 3.11)', matrix.test-case) }}
+    steps:
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.11"
+
+      - name: Azure login
+        uses: azure/login@v2
+        with:
+          client-id: ${{secrets.APICEXT_TEST_AZURE_CLIENT_ID}}
+          tenant-id: ${{secrets.APICEXT_TEST_AZURE_TENANT_ID}}
+          subscription-id: ${{secrets.APICEXT_TEST_AZURE_SUBSCRIPTION_ID}}
+          enable-AzPSSession: true
+
+      - name: Clone azure-cli-extensions repository
+        run: |
+          git clone https://github.com/blackchoey/azure-cli-extensions.git
+          cd azure-cli-extensions
+          git checkout main
+
+      - name: Setup development environment
+        run: |
+          cd azure-cli-extensions
+          python -m venv .venv
+          source .venv/bin/activate
+          python -m pip install -U pip
+          pip install azdev
+          azdev setup -r .
+          azdev extension add apic-extension
+
+      - name: Run tests
+        run: |
+          cd azure-cli-extensions
+          source .venv/bin/activate
+          if [ "${{ matrix.test-case }}" = "full-suite" ]; then
+            echo "Running full test suite..."
+            azdev test apic-extension --discover --live
+          else
+            echo "Running individual test case: ${{ matrix.test-case }}"
+            azdev test azext_apic-extension ${{ matrix.test-case }} --discover --live
+          fi
+        env:
+          USERASSIGNED_IDENTITY: ${{ secrets.USERASSIGNED_IDENTITY }}
+          AWS_ACCESS_KEY_LINK: ${{ secrets.AWS_ACCESS_KEY_LINK }}
+          AWS_SECRET_ACCESS_KEY_LINK: ${{ secrets.AWS_SECRET_ACCESS_KEY_LINK }}
+
+  execute-test-case-py312:
```

Comment:
are we able to use matrix or other github action features to run in different Python versions? So we don't need to duplicate the same logic multiple times.

Reply:
no we can't. There are 124 test cases currently. If using matrix for all 4 Python version, the matrix will be 124*4>256, where 256 is the max limit of the matrix in github action.
So I have to use 4 different job to do this.

Reply:
I found that there is an open-souce github action extension to handle matrix limitation. https://github.com/marketplace/actions/matrix-extended
But I think that maybe makes things much more complex.